### ISSUE TITLE
IMGT Convert function shows less sequences

### ISSUE THREAD

**Issue by SolaimanR** [mikessh/vdjtools#63]
Hey, 

we want to normalise our samples by only using X Sequences for each sample during analysis. We want to filter these X Sequences from the 3_Nt* file from IMGT, so right before we use "Convert" in VDJTOOLS. For this we filter all productive sequences from IMGT and use this file to get X Sequences. Than we run the "Convert" function in VDJTOOLS on that file. We have noticed, that the file contains less sequences than what we input. Why is that? So for instance, if we use 20000 productive sequences and use VDJTOOLS to parse them with "Convert", we get maybe 7000 or 3000 total number of sequences. I dont mean clonotypes, I am aware they will be less, but I mean the sum of the "count" column.

**Comment by mikessh** [comment]
There is a summary output for dropped records printed when running the ``Convert`` routine to the console, can you please share it. Note that IMGT output is really hard to parse so if you have raw data I would recommend to use other software (MIXCR, MIGMAP, etc)

**Comment by SolaimanR** [comment]
Hey. When I run 2 samples I get following output printed:

```
[Wed Mar 01 00:47:55 CET 2017 Convert] Reading sample(s)
[Wed Mar 01 00:47:55 CET 2017 Convert] 3 sample(s) loaded
[Wed Mar 01 00:47:55 CET 2017 SampleStreamConnection] Loading sample METADATA_FILENAME 
[WARNING] Some of the essential fields are bad/missing for the following clonotype string (displaying first 5 warnings)
BAD_LINE:
profolder_xprofolder_LK388.txz
BAD_LINE:
profolder_xprofolder_LK389.txz
[Wed Mar 01 00:47:55 CET 2017 ClonotypeStreamParser] Finished parsing. 1 header and 2 bad line(s) were skipped.
[Wed Mar 01 00:47:55 CET 2017 SampleStreamConnection] Loaded sample METADATA_FILENAME  with 0 clonotypes and 0 cells. Memory usage: 0 of 6 GB
[Wed Mar 01 00:47:55 CET 2017 Convert] Processed 1 sample(s).. Writing output
[Wed Mar 01 00:47:55 CET 2017 SampleStreamConnection] Loading sample profolder_xprofolder_LK388.txz
[WARNING] Some of the essential fields are bad/missing for the following clonotype string (displaying first 5 warnings)
BAD_LINE:
1	M01340:5:000000000-AUCK3:1:1101:16323:1693 1:N:0:6	No results	
BAD_LINE:
2	M01340:5:000000000-AUCK3:1:1101:20641:1697 1:N:0:6	No results	
BAD_LINE:
3	M01340:5:000000000-AUCK3:1:1101:22992:1700 1:N:0:6	No results	
BAD_LINE:
5	M01340:5:000000000-AUCK3:1:1101:11697:1717 1:N:0:6	No results	
BAD_LINE:
6	M01340:5:000000000-AUCK3:1:1101:13241:1720 1:N:0:6	No results	
[Wed Mar 01 00:47:56 CET 2017 ClonotypeStreamParser] Finished parsing. 1 header and 14871 bad line(s) were skipped.
[Wed Mar 01 00:47:56 CET 2017 SampleStreamConnection] Loaded sample profolder_xprofolder_LK388.txz with 2078 clonotypes and 5129 cells. Memory usage: 0 of 6 GB
[Wed Mar 01 00:47:56 CET 2017 Convert] Processed 2 sample(s).. Writing output
[Wed Mar 01 00:47:56 CET 2017 SampleStreamConnection] Loading sample profolder_xprofolder_LK389.txz
[WARNING] Some of the essential fields are bad/missing for the following clonotype string (displaying first 5 warnings)
BAD_LINE:
1	M01340:5:000000000-AUCK3:1:1101:21036:1691_1:N:0:4	No results	
BAD_LINE:
3	M01340:5:000000000-AUCK3:1:1101:11381:1730_1:N:0:4	No results	
BAD_LINE:
4	M01340:5:000000000-AUCK3:1:1101:17798:1733_1:N:0:4	No results	
BAD_LINE:
5	M01340:5:000000000-AUCK3:1:1101:8807:1746 1:N:0:46	No results	
BAD_LINE:
6	M01340:5:000000000-AUCK3:1:1101:9076:1791 1:N:0:46	No results	
[Wed Mar 01 00:47:57 CET 2017 ClonotypeStreamParser] Finished parsing. 1 header and 14416 bad line(s) were skipped.
[Wed Mar 01 00:47:57 CET 2017 SampleStreamConnection] Loaded sample profolder_xprofolder_LK389.txz with 2457 clonotypes and 5584 cells. Memory usage: 0 of 6 GB
[Wed Mar 01 00:47:57 CET 2017 Convert] Processed 3 sample(s).. Writing output
[Wed Mar 01 00:47:57 CET 2017 Convert] Finished
```

**Comment by mikessh** [comment]
This tells that the command line arguments were incorrect - ``METADATA_FILENAME`` which doesn't contain any clonotypes was somehow passed as an input file, likely you've omitted ``-m METADATA_FILENAME``.

**Comment by SolaimanR** [comment]
Hey, what is strange though, is that I always get my parsed files. I havent imitted -m METADATA_FILENAME its in the command. So even if METADATA_FILENAME is passed as a input file, it seems to still parse all other files correctly. However, with fewer reads than when I manually sort out all productive reads.

What could be the problem? What does this line tell you?:

`[WARNING] Some of the essential fields are bad/missing for the following clonotype string (displaying first 5 warnings)
BAD_LINE:
1	M01340:5:000000000-AUCK3:1:1101:21036:1691_1:N:0:4	No results	
BAD_LINE:
3	M01340:5:000000000-AUCK3:1:1101:11381:1730_1:N:0:4	No results	
BAD_LINE:
4	M01340:5:000000000-AUCK3:1:1101:17798:1733_1:N:0:4	No results	
BAD_LINE:
5	M01340:5:000000000-AUCK3:1:1101:8807:1746 1:N:0:46	No results	
BAD_LINE:
6	M01340:5:000000000-AUCK3:1:1101:9076:1791 1:N:0:46	No results`

**Comment by SolaimanR** [comment]
I found out why METADATA_FILENAME is being passed as input and fixed that. However, I still have sequences missing. I still get this warning, and it specifically says "X number of bad lines were skipped". Im assuming those are the sequences missing, why are they considered "bad lines"?

`[Tue Mar 07 07:19:06 CET 2017 Convert] Reading sample(s)
[Tue Mar 07 07:19:06 CET 2017 Convert] 2 sample(s) loaded
[Tue Mar 07 07:19:06 CET 2017 SampleStreamConnection] Loading sample profolder_LK388.txz
[WARNING] Some of the essential fields are bad/missing for the following clonotype string (displaying first 5 warnings)
BAD_LINE:
1	M01340:5:000000000-AUCK3:1:1101:16323:1693 1:N:0:6	No results	
BAD_LINE:
2	M01340:5:000000000-AUCK3:1:1101:20641:1697 1:N:0:6	No results	
BAD_LINE:
3	M01340:5:000000000-AUCK3:1:1101:22992:1700 1:N:0:6	No results	
BAD_LINE:
5	M01340:5:000000000-AUCK3:1:1101:11697:1717 1:N:0:6	No results	
BAD_LINE:
6	M01340:5:000000000-AUCK3:1:1101:13241:1720 1:N:0:6	No results	
[Tue Mar 07 07:19:09 CET 2017 ClonotypeStreamParser] Finished parsing. 1 header and 84742 bad line(s) were skipped.
[Tue Mar 07 07:19:09 CET 2017 SampleStreamConnection] Loaded sample profolder_LK388.txz with 10735 clonotypes and 33271 cells. Memory usage: 0 of 6 GB
[Tue Mar 07 07:19:09 CET 2017 Convert] Processed 1 sample(s).. Writing output
[Tue Mar 07 07:19:09 CET 2017 SampleStreamConnection] Loading sample profolder_LK389.txz
[WARNING] Some of the essential fields are bad/missing for the following clonotype string (displaying first 5 warnings)
BAD_LINE:
1	M01340:5:000000000-AUCK3:1:1101:21036:1691_1:N:0:4	No results	
BAD_LINE:
3	M01340:5:000000000-AUCK3:1:1101:11381:1730_1:N:0:4	No results	
BAD_LINE:
4	M01340:5:000000000-AUCK3:1:1101:17798:1733_1:N:0:4	No results	
BAD_LINE:
5	M01340:5:000000000-AUCK3:1:1101:8807:1746 1:N:0:46	No results	
BAD_LINE:
6	M01340:5:000000000-AUCK3:1:1101:9076:1791 1:N:0:46	No results	
[Tue Mar 07 07:19:11 CET 2017 ClonotypeStreamParser] Finished parsing. 1 header and 53815 bad line(s) were skipped.
`

**Comment by SolaimanR** [comment]
Hey, could you look into this? We still cant figure out what the "X numer of bad lines" are and how to prevent them from being skipped.

**Comment by mikessh** [comment]
Well it looks that your bed lines are read headers (M01340:5:000000000-AUCK3:1:1101:9076:1791 1:N:0:46). VDJtools is supposed to work with processed clonotype files, so its likely the wrong file from IMGT. You can see the example IMGT file used in tests here: https://github.com/mikessh/vdjtools/tree/master/src/test/resources/samples

Anyways you can try any other VDJ mapping software from the list of software tools supported by VDJtools..

**Comment by SolaimanR** [comment]
Is the sample file downloaded from IMGT? I cant open it or extract it.

**Comment by SolaimanR** [comment]
I checked your sample file. Thats how ours looks like. The only difference are the Sequence IDs, that of course are different. However, all our samples have the sequence/read ID format and only X numbers are skipped.

**Comment by SolaimanR** [comment]
Ok mikessh, I figured out half the problem. Our IMGT files had, in addition to "productive" and "unproductive" also "No results", which caused most of the bad lines. I removed the "No results" lines, but there are still bad lines now. 

`M01340:5:000000000-AUCK3:1:1101:17089:2711 1:N:0:6	unknown	Homsap TRBV20-1*01 F, or Homsap TRBV20-1*02 F or Homsap TRBV20-1*03 (F) or Homsap TRBV20-1*04 (F) or Homsap TRBV20-1*05 (F) or Homsap TRBV20-1*06 (F) or Homsap TRBV20-1*07 (F)	gtgaccagggccgaggatgaagagagcagcgggtacatatga					gtgaccagggccgaggatgaagagagcagcgggtacatatga	142									1	42						
BAD_LINE:
173	M01340:5:000000000-AUCK3:1:1101:7491:3331 1:N:0:6	unknown	Homsap TRBV5-3*01 ORF, or Homsap TRBV5-3*02 ORF		gaggagtgagtcacccaaaaggccacactggtgtgc	gaggagtgagtcacccaaaaggccacactggtgtgc							165	200	165	200													
BAD_LINE:
202	M01340:5:000000000-AUCK3:1:1101:24527:3543 1:N:0:6	unknown	Homsap TRBV6-1*01 F					tggattcggctgctccctcccagacatctgtgtacttctgtgcca					tggattcggctgctccctcccagacatctgtgtacttctgt	gcca		tgtgcca														145									1	41	42	45			39	45																3
BAD_LINE:
246	M01340:5:000000000-AUCK3:1:1101:12797:3999 1:N:0:6	unknown	Homsap TRBV5-1*01 F, or Homsap TRBV5-1*02 (F)		gcacctttgagtttgggtattcagcctctgatcttcgcgccagc					gcacctttgagtttgggtattcagcctctgatcttcgc	gccagccgcgccagc															144									1	38	39	44			36	44																3
BAD_LINE:
273	M01340:5:000000000-AUCK3:1:1101:22444:4245 1:N:0:6	unknown	Homsap TRBV27*01 F					agatgaaggactggtgagttgacttatgagggagatgccatcgaagggtgtaaagttcctataatagaccagtcgactttcgccccgggaccagggtcaccgtgccccaggaactcctctatgtactcctgc			aga	tgaaggactggtgagttg	acttatgagggagatgccatcgaagggtgtaaagttcctataatagaccagtcgactttcgccccgggaccagggtcaccgtgccccaggaactcctctatgtactcctgc											1132					1	3	4	21	22	132						1
[Sat May 06 11:25:59 CEST 2017 ClonotypeStreamParser] Finished parsing. 1 header and 1183 bad line(s) were skipped.
`

I am looking into this. I hope its okay if I post the updates here, I think IMGT has changed some of there data structure and it might be useful for others to follow with this. Maybe you have an idea what it might be.
